### PR TITLE
CDAP-15824 File-appender plugin validation

### PR DESCRIPTION
**Pull request by mlozbin-cybervisiontech**
JIRA: https://issues.cask.co/browse/CDAP-15824
WIKI: https://wiki.cask.co/display/CE/Guidelines+on+Plugin+Validation

In scope of this PR:
- integrated with new validation API

### CODE REVIEW

**Comment by mlozbin-cybervisiontech** [comment]
@albertshau could you, please, take a look.

File: src/test/java/io/cdap/plugin/fileappend/FileAppenderSinkConfigTest.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    List<ValidationFailure> failureList = failureCollector.getValidationFailures();
+    Assert.assertEquals(1, failureList.size());
+    ValidationFailure failure = failureList.get(0);
+    List<ValidationFailure.Cause> causeList = failure.getCauses()
+      .stream()
+      .filter(cause -> cause.getAttribute(CauseAttributes.STACKTRACE) != null)
+      .collect(Collectors.toList());
+    Assert.assertEquals(1, causeList.size());
+  }
+}
```
Comment (CuriousVini) [inline src/test/java/io/cdap/plugin/fileappend/FileAppenderSinkConfigTest.java]: nit: new line

File: src/main/java/io/cdap/plugin/fileappend/FileAppenderSinkConfig.java [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+  public String getSchema() {
+    return schema;
+  }
+
+  @Nullable
+  public Schema getParsedSchema() {
+    try {
+      return schema != null ? Schema.parseJson(schema) : null;
+    } catch (IOException e) {
+      throw new IllegalArgumentException("Could not parse schema: " + e.getMessage(), e);
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/fileappend/FileAppenderSinkConfig.java]: to keep the error message consistent change this to `Invalid schema :  + e.getMessage()` and where it is caught just use `e.getMessage()` as error message

File: src/main/java/io/cdap/plugin/fileappend/FileAppenderSinkConfig.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    } catch (IllegalArgumentException e) {
+      failureCollector.addFailure("Invalid output schema: " + e.getMessage(), null)
+        .withConfigProperty(SCHEMA)
+        .withStacktrace(e.getStackTrace());
+    }
+    if (inputSchema != null && parsedSchema != null) {
+      for (Schema.Field outputField : parsedSchema.getFields()) {
+        Schema.Field inField = inputSchema.getField(outputField.getName());
+        if (inField == null) {
+          failureCollector.addFailure("Field is not in the input schema.", null)
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/fileappend/FileAppenderSinkConfig.java]: `Field %s must be present in the input schema.`

File: src/main/java/io/cdap/plugin/fileappend/FileAppenderSinkConfig.java [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+            .withOutputSchemaField(outputField.getName());
+        } else if (!inField.getSchema().equals(outputField.getSchema())) {
+          failureCollector.addFailure("Schema for field does not match the schema of the input field", null)
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/fileappend/FileAppenderSinkConfig.java]: message -`Field %s has type %s that does not match with input field schema %s`.   - use `schema.getDisplayName()` to get type of the field
corrective action - `Ensure output schema of %s matches input schema`

File: src/main/java/io/cdap/plugin/fileappend/FileAppenderSinkConfig.java [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+            .withOutputSchemaField(outputField.getName());
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/fileappend/FileAppenderSinkConfig.java:114]: `.withOutputSchemaField(outputField.getName()).withInputSchemaField(outputField.getName())`
we should highlight both input and output schema fields for this error.

File: src/main/java/io/cdap/plugin/fileappend/FileAppenderSink.java [hunk 6]
```diff
@@ -53,131 +51,71 @@
...
     // TODO: validate input schema
     Schema inputSchema = pipelineConfigurer.getStageConfigurer().getInputSchema();
     Schema outputSchema = inputSchema;
-    if (config.schema != null) {
-      try {
-        outputSchema = Schema.parseJson(config.schema);
-      } catch (IOException e) {
-        throw new IllegalArgumentException("Could not parse schema: " + e.getMessage(), e);
-      }
+    Schema configuredSchema = config.getParsedSchema();
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/fileappend/FileAppenderSink.java:71]: This would throw `IllegalArgumentException` which will terminate the control.

File: src/main/java/io/cdap/plugin/fileappend/FileAppenderSink.java [hunk 7]
```diff
@@ -53,131 +51,71 @@
...
-        }
-      }
-
       SchemaConverter schemaConverter = new SchemaConverter(true);
       try {
         hiveSchema = schemaConverter.toHiveSchema(outputSchema);
       } catch (UnsupportedTypeException e) {
-        throw new IllegalArgumentException("Could not generate Hive schema: " + e.getMessage(), e);
+        failureCollector.addFailure("Could not generate Hive schema: " + e.getMessage(), null)
+          .withStacktrace(e.getStackTrace());
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/fileappend/FileAppenderSink.java:84]: we can add another cause `.withConfigProperty("schema")`

**Review by CuriousVini — CHANGES_REQUESTED** [review]

File: src/main/java/io/cdap/plugin/fileappend/FileAppenderSink.java (hunk 6 above)
Comment (CuriousVini) [reply src/main/java/io/cdap/plugin/fileappend/FileAppenderSink.java:71]: Unaddressed?

Comment (mlozbin-cybervisiontech) [reply src/main/java/io/cdap/plugin/fileappend/FileAppenderSink.java:71]: addressed in last commit

**Comment by mlozbin-cybervisiontech** [comment]
@albertshau @CuriousVini could you, please, take a look.

File: widgets/FileAppender-batchsink.json [hunk 8]
```diff
@@ -15,7 +15,10 @@
         {
           "widget-type": "textbox",
           "label": "Field separator",
-          "name": "fieldSeparator"
+          "name": "fieldSeparator",
+          "widget-attributes": {
+            "default": ","
```
Comment (CuriousVini) [inline widgets/FileAppender-batchsink.json:20]: why is this change needed now?

File: src/main/java/io/cdap/plugin/fileappend/FileAppenderSink.java [hunk 9]
```diff
@@ -53,131 +51,76 @@
...
-        outputSchema = Schema.parseJson(config.schema);
-      } catch (IOException e) {
-        throw new IllegalArgumentException("Could not parse schema: " + e.getMessage(), e);
+    try {
+      Schema configuredSchema = config.getParsedSchema();
+      if (configuredSchema != null) {
+        outputSchema = configuredSchema;
       }
+    } catch (IllegalArgumentException e) {
+      failureCollector.getOrThrowException();
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/fileappend/FileAppenderSink.java]: this should add the schema failure to failure collector

File: src/main/java/io/cdap/plugin/fileappend/FileAppenderSink.java [hunk 10]
```diff
@@ -53,131 +51,76 @@
...
+  public FileAppenderSink(FileAppenderSinkConfig config) {
     this.config = config;
   }
 
   @Override
   public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
-    config.validate();
+    FailureCollector failureCollector = pipelineConfigurer.getStageConfigurer().getFailureCollector();
+    config.validate(failureCollector);
     // TODO: validate input schema
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/fileappend/FileAppenderSink.java]: lets add validation for input schema as well.

File: src/main/java/io/cdap/plugin/fileappend/FileAppenderSinkConfig.java [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+        if (inField == null) {
+          failureCollector.addFailure(String.format("Field '%s' must be present in the input schema.",
+                                                    outputField.getName()), null)
+            .withOutputSchemaField(outputField.getName());
+        } else if (!inField.getSchema().equals(outputField.getSchema())) {
+          failureCollector.addFailure(
+            String.format("Field '%s' has type '%s' that does not match with input field schema '%s'",
+                          outputField.getName(), outputField.getSchema().getDisplayName(),
+                          inField.getSchema().getDisplayName()),
+            String.format("Ensure output schema of '%s' matches input schema", outputField.getName()))
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/fileappend/FileAppenderSinkConfig.java]: should add period to the end of error message and corrective action

File: src/main/java/io/cdap/plugin/fileappend/FileAppenderSinkConfig.java [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+        .withStacktrace(e.getStackTrace());
+    }
+    if (inputSchema != null && parsedSchema != null) {
+      for (Schema.Field outputField : parsedSchema.getFields()) {
+        Schema.Field inField = inputSchema.getField(outputField.getName());
+        if (inField == null) {
+          failureCollector.addFailure(String.format("Field '%s' must be present in the input schema.",
+                                                    outputField.getName()), null)
+            .withOutputSchemaField(outputField.getName());
+        } else if (!inField.getSchema().equals(outputField.getSchema())) {
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/fileappend/FileAppenderSinkConfig.java]: we should compare nonnullable schema similar to https://github.com/data-integrations/google-cloud/blob/c39c7a2cbd395d4766abe3762b8496c4222f2466/src/main/java/io/cdap/plugin/gcp/bigtable/source/BigtableSourceConfig.java#L233

File: src/main/java/io/cdap/plugin/fileappend/FileAppenderSinkConfig.java [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+          failureCollector.addFailure(
+            String.format("Field '%s' has type '%s' that does not match with input field schema '%s'",
+                          outputField.getName(), outputField.getSchema().getDisplayName(),
+                          inField.getSchema().getDisplayName()),
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/fileappend/FileAppenderSinkConfig.java]: use `getDisplayName` for non-nullable schema

File: src/main/java/io/cdap/plugin/fileappend/FileAppenderSink.java (hunk 9 above)
Comment (CuriousVini) [reply src/main/java/io/cdap/plugin/fileappend/FileAppenderSink.java]: oh is it because we have added that failure in `config.validateOutputSchema`? If so, add that as a comment here.

File: widgets/FileAppender-batchsink.json (hunk 8 above)
Comment (mlozbin-cybervisiontech) [reply widgets/FileAppender-batchsink.json:20]: 1. to make it explicit to end user
2. to make it similar to other plugins

**Comment by mlozbin-cybervisiontech** [comment]
@CuriousVini addressed all recent comments

File: src/main/java/io/cdap/plugin/fileappend/FileAppenderSink.java (hunk 10 above)
Comment (mlozbin-cybervisiontech) [reply src/main/java/io/cdap/plugin/fileappend/FileAppenderSink.java]: I'm not sure what should be validated in input schema. Do you have any requirements? As I can see, existing validation may be enough in most cases.

**Comment by CuriousVini** [comment]
Is this missed? https://github.com/data-integrations/file-appender/pull/8/files#r324951676

File: src/main/java/io/cdap/plugin/fileappend/FileAppenderSinkConfig.java [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+
+    private Builder() {
+    }
+
+    public Builder setName(String val) {
+      name = val;
+      return this;
+    }
+
+    public Builder setOutputDir(String val) {
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/fileappend/FileAppenderSinkConfig.java]: this is nullable. Mark all the fields that are expected to be nullable

File: src/main/java/io/cdap/plugin/fileappend/FileAppenderSinkConfig.java [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static Builder builder(FileAppenderSinkConfig copy) {
+    return new Builder()
+      .setName(copy.getName())
+      .setOutputDir(copy.getOutputDir())
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/fileappend/FileAppenderSinkConfig.java]: what would happen when output directory is null or empty? since its an optional field

Comment (CuriousVini) [reply src/main/java/io/cdap/plugin/fileappend/FileAppenderSinkConfig.java]: similar question for other fields as well.

**Review by CuriousVini — COMMENTED** [review]
minor comment about marking fields as nullable. Validation logic looks good.

**Review by CuriousVini — APPROVED** [review]
please squash after addressing last comment

**Comment by CuriousVini** [comment]
please address the comments and squash the commits, thanks!
